src/NuGet.Core/NuGet.Build.Tasks/StaticGraphRestoreTaskBase.cs

```diff
@@ -0,0 +1,273 @@
+
+        internal readonly CancellationTokenSource _cancellationTokenSource = new CancellationTokenSource();
+
+        /// <summary>
+        /// Gets a value indicating whether or not <see cref="SolutionPath" /> contains a value.
+        /// </summary>
+        public bool IsSolutionPathDefined => !string.IsNullOrWhiteSpace(SolutionPath) && !string.Equals(SolutionPath, "*Undefined*", StringComparison.OrdinalIgnoreCase);
+
+        /// <summary>
+        /// Gets or sets the full path to the directory containing MSBuild.
+        /// </summary>
+        [Required]
+        public string MSBuildBinPath { get; set; }
+
+        /// <summary>
+        /// MSBuildStartupDirectory - Used to calculate relative paths
+        /// </summary>
+        public string MSBuildStartupDirectory { get; set; }
+
+        /// <summary>
+        /// The path to the file to start the additional process with.
+        /// </summary>
+        public string ProcessFileName { get; set; }
+
+        /// <summary>
+        /// Gets or sets the full path to the current project file.
+        /// </summary>
+        [Required]
+        public string ProjectFullPath { get; set; }
+
+        /// <summary>
+        /// Get or sets a value indicating whether or not the restore should restore all projects or just the entry project.
+        /// </summary>
+        public bool Recursive { get; set; }
+
+        /// <summary>
+        /// Gets or sets the full path to the solution file (if any) that is being built.
+        /// </summary>
+        public string SolutionPath { get; set; }
+
+        protected abstract string DebugEnvironmentVariableName { get; }
+
+        /// <inheritdoc cref="ICancelableTask.Cancel" />
+        public void Cancel() => _cancellationTokenSource.Cancel();
+
+        /// <inheritdoc cref="IDisposable.Dispose" />
+        public void Dispose()
+        {
+            Dispose(disposing: true);
+
+            GC.SuppressFinalize(this);
+        }
+
+        public override bool Execute()
+        {
+            try
+            {
+#if DEBUG
+                var debugRestoreTask = Environment.GetEnvironmentVariable(DebugEnvironmentVariableName);
+                if (!string.IsNullOrEmpty(debugRestoreTask) && debugRestoreTask.Equals(bool.TrueString, StringComparison.OrdinalIgnoreCase))
+                {
+                    Debugger.Launch();
+                }
+#endif
+                MSBuildLogger logger = new MSBuildLogger(Log);
+
+                using (var semaphore = new SemaphoreSlim(initialCount: 0, maxCount: 1))
+                using (var loggingQueue = new TaskLoggingQueue(Log))
+                using (var process = new Process())
+                {
+                    process.EnableRaisingEvents = true;
+                    process.StartInfo = new ProcessStartInfo
+                    {
+                        Arguments = $"\"{string.Join("\" \"", GetCommandLineArguments(MSBuildBinPath))}\"",
```

Comment:
Given your micro-optimizations of other code, I'm a little surprised you're using both interpolated strings and `string.Join` here.  I don't know if it'll add up to even a full millisecond of improvement, it might be a lot less. But, since `GetCommandLineArguments` always returns the same number of arguments, it is theoretically faster to have everything in a single `string.Concat` (or use the `str1 + str2` syntax).

Reply:
The primary reason I like `string.Join()` in this case is it ensures that each argument is surrounded by quotes and the `GetCommandLineArguments()` method just looks so clean to me returning `IEnumerable<string>`.  `GetCommandLineArguments()` also has to return different values for .NET Core vs .NET Framework so I was afraid of just hardcoding everything.

This just _looks_ terrible to me:

```c#
internal string GetCommandLineArguments(string msbuildBinPath)
{
    return
#if IS_CORECLR
    "\"" + Path.Combine(ThisAssemblyLazy.Value.DirectoryName, Path.ChangeExtension(ThisAssemblyLazy.Value.Name, ".Console.dll")) + "\""
        + " \"" + Path.Combine(msbuildBinPath, "MSBuild.dll") + "\""
#else
    "\"" + Path.Combine(msbuildBinPath, "MSBuild.exe") + "\""
#endif
    + " \"" + (IsSolutionPathDefined ? SolutionPath : ProjectFullPath) + "\"";
}
```

Should I do that instead?

Reply:
The perf difference isn't anywhere near as much as I would have expected.

<details>
<summary>code</summary>
<p>

```cs
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

[MemoryDiagnoser]
[SimpleJob(BenchmarkDotNet.Jobs.RuntimeMoniker.Net70)]
[SimpleJob(BenchmarkDotNet.Jobs.RuntimeMoniker.Net472)]
public class Program
{
    private string MSBuildBinPath => string.Empty;
    private bool IsSolutionPathDefined => true;
    private string SolutionPath => string.Empty;
    private string ProjectFullPath => string.Empty;

    private static readonly Lazy<FileInfo> ThisAssemblyLazy = new Lazy<FileInfo>(() => new FileInfo(typeof(Program).Assembly.Location));

    static void Main()
    {
        BenchmarkRunner.Run<Program>();
    }

    [Benchmark]
    public string StringJoinIEnumerable()
    {
        return $"\"{string.Join("\" \"", GetCommandLineArgumentsIEnumerable(MSBuildBinPath))}\"";
    }

    internal IEnumerable<string> GetCommandLineArgumentsIEnumerable(string msbuildBinPath)
    {
#if !NETFRAMEWORK
        yield return Path.Combine(ThisAssemblyLazy.Value.DirectoryName, Path.ChangeExtension(ThisAssemblyLazy.Value.Name, ".Console.dll"));
        yield return Path.Combine(msbuildBinPath, "MSBuild.dll");
#else
        yield return Path.Combine(msbuildBinPath, "MSBuild.exe");
#endif
        yield return IsSolutionPathDefined ? SolutionPath : ProjectFullPath;
    }

    [Benchmark]
    public string StringJoinArray()
    {
        return $"\"{string.Join("\" \"", GetCommandLineArgumentsArray(MSBuildBinPath))}\"";
    }

    internal string[] GetCommandLineArgumentsArray(string msbuildBinPath)
    {
        return new [] {

#if !NETFRAMEWORK
        Path.Combine(ThisAssemblyLazy.Value.DirectoryName, Path.ChangeExtension(ThisAssemblyLazy.Value.Name, ".Console.dll")),
        Path.Combine(msbuildBinPath, "MSBuild.dll"),
#else
        Path.Combine(msbuildBinPath, "MSBuild.exe"),
#endif
        IsSolutionPathDefined ? SolutionPath : ProjectFullPath
        };
    }

    [Benchmark]
    public string Concat()
    {
        return
        #if !NETFRAMEWORK 
        "\"" + Path.Combine(ThisAssemblyLazy.Value.DirectoryName, Path.ChangeExtension(ThisAssemblyLazy.Value.Name, ".Console.dll")) + "\"" + " \"" + Path.Combine(MSBuildBinPath, "MSBuild.dll") + "\""
        #else 
        "\"" + Path.Combine(MSBuildBinPath, "MSBuild.exe") + "\"" 
        #endif 
        + " \"" + (IsSolutionPathDefined ? SolutionPath : ProjectFullPath) + "\"";
    }
}
```

</p>
</details>

on my machine, the results are:

|                Method |                  Job |              Runtime |     Mean |   Error |  StdDev |   Gen0 |   Gen1 | Allocated |
|---------------------- |--------------------- |--------------------- |---------:|--------:|--------:|-------:|-------:|----------:|
| StringJoinIEnumerable |             .NET 7.0 |             .NET 7.0 | 210.6 ns | 1.65 ns | 1.38 ns | 0.1442 | 0.0002 |    1208 B |
|       StringJoinArray |             .NET 7.0 |             .NET 7.0 | 179.2 ns | 0.94 ns | 0.84 ns | 0.1433 | 0.0002 |    1200 B |
|                Concat |             .NET 7.0 |             .NET 7.0 | 178.9 ns | 1.32 ns | 1.03 ns | 0.1118 | 0.0002 |     936 B |
| StringJoinIEnumerable | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 255.0 ns | 0.62 ns | 0.55 ns | 0.0277 |      - |     177 B |
|       StringJoinArray | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 205.9 ns | 0.61 ns | 0.54 ns | 0.0253 |      - |     160 B |
|                Concat | .NET Framework 4.7.2 | .NET Framework 4.7.2 | 208.9 ns | 0.98 ns | 0.86 ns | 0.0305 |      - |     193 B |

So, the perf gains are insignificant, particularly since this is code that runs once per process. You used perf as a justification for other changes, like writing a custom JsonReader/Writer code, though in that case the perf gain was much higher (though you didn't provide evidence).  However, the benchmark does show that using IEnumerable and `yield return` has a perf impact (significant in percentage, insignificant in absolute terms). I guess the lesson here is that your perf instincts are much better than mine, so I was surprised that you wanted to gain ~1ms from adding a bunch of tech debt in json (de)serialization code, but then use keep using other non-ideal patterns (which really only matter in tight loops, or high scale web apps)

Reply:
Thanks for the benchmarks!

> You used perf as a justification for other changes, like writing a custom JsonReader/Writer code, though in that case the perf gain was much higher (though you didn't provide evidence).

JSON deserialization has a lot of overhead because of type checking via reflection.  Its really handy to declare a class and get an instance back from JSON but in my experience its just not fast.  Our code base seems to generally lean towards custom deserialization/JSON reading:

https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Packaging/NupkgMetadata/NupkgMetadataFileFormat.cs#L127
https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Packaging/RuntimeModel/JsonRuntimeFormat.cs#L43
https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs#L239

This helped me make the decision to use custom JSON reading and thought it was a well established pattern not technical debt.  We also spoke over chat about maybe not even using JSON now that its writing to StandardInput.  The only thing I absolutely have to send over IPC is the global properties and options which are dictionaries.  I could explore those further as well.  Maybe I can use `Span<T>` to eke out every last bit of perf? 

As for calling `string.Join()` on an `IEnumerable<string>` I knew it was going to allocate more memory but was assuming it would be very fast since it's not doing anything like type checking when deserializing JSON.  I felt this was still performant at the cost of memory which at the point this part of the code runs there shouldn't be any GC pressure since its running pretty early in `MSBuild.exe`.  

The bytes allocated seem off, I'd expect Concat to be way lower?  I'm fine with using it, I guess I just wish C# had a better way of doing it.

Reply:
I've settled on the `string.Concat()` syntax after I read that `" " + " "` gets turned into that by the compiler and it looks a lot cleaner to me.  Thanks for the suggestion!